src/adapter/src/error.rs

```diff
@@ -604,6 +615,61 @@ impl AdapterError {
     pub fn internal<E: std::fmt::Display>(context: &str, e: E) -> AdapterError {
         AdapterError::Internal(format!("{context}: {e}"))
     }
+
+    // We don't want the following error conversions to `ConcurrentDependencyDrop` to happen
+    // automatically, because it might depend on the context whether `ConcurrentDependencyDrop`
+    // is appropriate, so we want to make the conversion target explicit at the call site.
+    // For example, maybe we get an `InstanceMissing` if the user specifies a non-existing cluster,
+    // in which case `ConcurrentDependencyDrop` would not be appropriate.
+    pub fn concurrent_dependency_drop_from_instance_missing(e: InstanceMissing) -> Self {
+        AdapterError::ConcurrentDependencyDrop {
+            dependency_kind: "cluster",
+            dependency_id: e.0.to_string(),
+        }
+    }
+    pub fn concurrent_dependency_drop_from_collection_missing(e: CollectionMissing) -> Self {
+        AdapterError::ConcurrentDependencyDrop {
+            dependency_kind: "collection",
```

Comment:
Is there any precedent for referring to "collections" in our SQL API? I think usually we refer to tables, sources, views, etc. and a collection is more of an internal concept. If we are not able to distinguish the item type then maybe calling it "item" is the right SQL thing to do?